File: src/topk.ts
import {
  type ProposalTrace,
  type Receipt,
  type TransactionEngine,
  type TypedCandidate,
  makeTrace,
} from "./core.js";
import { type ResidualSignal, ResidualTaxonomyMemory } from "./residuals.js";

export interface ResidualRepairOption<CandidatePayload> {
  label: string;
  candidate: TypedCandidate<CandidatePayload>;
  repairHint: string;
  baseRank: number;
}

export interface ResidualTopKOutcome<State> {
  state: State;
  committed: boolean;
  committedLabel: string;
  submittedLabels: string[];
  receipts: Receipt[];
  verifierCalls: number;
  reason: "commit" | "top_k_exhausted";
}

export class ResidualTopKSubmitter<State, CandidatePayload> {
  engine: TransactionEngine<State, CandidatePayload>;
  memory: ResidualTaxonomyMemory;

  constructor(engine: TransactionEngine<State, CandidatePayload>, memory = new ResidualTaxonomyMemory()) {
    this.engine = engine;
    this.memory = memory;
  }

  rankOptions(
    options: Array<ResidualRepairOption<CandidatePayload>>,
    residualSignal: ResidualSignal | null = null,
  ): Array<ResidualRepairOption<CandidatePayload>> {
    const preferred = this.preferredHints(residualSignal);
    return [...options].sort((a, b) => {
      const aPreferred = a.repairHint && preferred.has(a.repairHint) ? 0 : 1;
      const bPreferred = b.repairHint && preferred.has(b.repairHint) ? 0 : 1;
      return aPreferred - bPreferred
        || a.baseRank - b.baseRank
        || compareCodePoint(a.label, b.label);
    });
  }

  async submit(
    state: State,
    options: Array<ResidualRepairOption<CandidatePayload>>,
    params: {
      topK: number;
      tracePrefix: string;
      residualSignal?: ResidualSignal | null;
      modelVersion?: string;
    },
  ): Promise<ResidualTopKOutcome<State>> {
    if (!Number.isInteger(params.topK) || params.topK < 0) {
      throw new RangeError("topK must be a non-negative integer");
    }
    const ranked = this.rankOptions(options, params.residualSignal ?? null);
    const receipts: Receipt[] = [];
    const submittedLabels: string[] = [];
    let current = state;
    for (let idx = 0; idx < Math.min(params.topK, ranked.length); idx += 1) {
      const option = ranked[idx];
      submittedLabels.push(option.label);
      const trace: ProposalTrace = makeTrace({
        branchId: `${params.tracePrefix}-${idx}-${option.label}`,
        actions: [{ label: option.label, repairHint: option.repairHint }],
        modelVersion: params.modelVersion ?? "residual.topk.v1",
      });
      const outcome = await this.engine.transact(state, trace, option.candidate);
      receipts.push(outcome.receipt);
      if (outcome.committed) {
        current = outcome.state;
        return {
          state: current,
          committed: true,
          committedLabel: option.label,
          submittedLabels,
          receipts,
          verifierCalls: receipts.length,
          reason: "commit",
        };
      }
    }
    return {
      state: current,
      committed: false,
      committedLabel: "",
      submittedLabels,
      receipts,
      verifierCalls: receipts.length,
      reason: "top_k_exhausted",
    };
  }

  private preferredHints(residualSignal: ResidualSignal | null): Set<string> {
    const preferred = new Set<string>();
    if (!residualSignal) {
      return preferred;
    }
    for (const hint of residualSignal.repairHints) {
      preferred.add(hint);
    }
    const topHint = this.memory.topRepairHint(residualSignal.kind);
    if (topHint) {
      preferred.add(topHint);
    }
    return preferred;
  }
}

function compareCodePoint(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

File: trwm/topk.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Iterable

from .core import ProposalTrace, Receipt, TransactionEngine, TypedCandidate
from .residuals import ResidualSignal, ResidualTaxonomyMemory


@dataclass(frozen=True)
class ResidualRepairOption:
    label: str
    candidate: TypedCandidate
    repair_hint: str = ""
    base_rank: int = 0


@dataclass(frozen=True)
class ResidualTopKOutcome:
    state: Any
    committed: bool
    committed_label: str
    submitted_labels: tuple[str, ...]
    receipts: tuple[Receipt, ...]
    verifier_calls: int
    reason: str


class ResidualTopKSubmitter:
    """Rank and submit bounded repair candidates; hard verifiers still own commits."""

    def __init__(self, engine: TransactionEngine, memory: ResidualTaxonomyMemory | None = None):
        self.engine = engine
        self.memory = memory or ResidualTaxonomyMemory()

    def rank_options(
        self,
        options: Iterable[ResidualRepairOption],
        *,
        residual_signal: ResidualSignal | None = None,
    ) -> tuple[ResidualRepairOption, ...]:
        rows = tuple(options)
        preferred = self._preferred_hints(residual_signal)
        return tuple(
            sorted(
                rows,
                key=lambda option: (
                    0 if option.repair_hint and option.repair_hint in preferred else 1,
                    option.base_rank,
                    option.label,
                ),
            )
        )

    def submit(
        self,
        state: Any,
        options: Iterable[ResidualRepairOption],
        *,
        top_k: int,
        trace_prefix: str,
        residual_signal: ResidualSignal | None = None,
        model_version: str = "residual.topk.v1",
    ) -> ResidualTopKOutcome:
        if not isinstance(top_k, int) or isinstance(top_k, bool) or top_k < 0:
            raise ValueError("top_k must be a non-negative integer")
        ranked = self.rank_options(options, residual_signal=residual_signal)
        receipts: list[Receipt] = []
        submitted: list[str] = []
        current = state
        for idx, option in enumerate(ranked[:top_k]):
            submitted.append(option.label)
            outcome = self.engine.transact(
                state,
                ProposalTrace(
                    branch_id=f"{trace_prefix}-{idx}-{option.label}",
                    actions=({"label": option.label, "repair_hint": option.repair_hint},),
                    model_version=model_version,
                ),
                option.candidate,
            )
            receipts.append(outcome.receipt)
            if outcome.committed:
                current = outcome.state
                return ResidualTopKOutcome(
                    state=current,
                    committed=True,
                    committed_label=option.label,
                    submitted_labels=tuple(submitted),
                    receipts=tuple(receipts),
                    verifier_calls=len(receipts),
                    reason="commit",
                )
        return ResidualTopKOutcome(
            state=current,
            committed=False,
            committed_label="",
            submitted_labels=tuple(submitted),
            receipts=tuple(receipts),
            verifier_calls=len(receipts),
            reason="top_k_exhausted",
        )

    def _preferred_hints(self, residual_signal: ResidualSignal | None) -> set[str]:
        preferred: set[str] = set()
        if residual_signal is not None:
            preferred.update(residual_signal.repair_hints)
            top_hint = self.memory.top_repair_hint(residual_signal.kind)
            if top_hint:
                preferred.add(top_hint)
        return preferred
